### ISSUE TITLE
[Bug] Ryujinx impl of LayeredFS does not match Atmosphere

### ISSUE THREAD

**Issue by hYdos** [Ryujinx/Ryujinx#5661]
### Description of the issue

I created a mod to try load the game faster by extracting the games internal archive and invalidating its entries so that it would try grab the files from the root RomFS folder and tried to run it and Ryujinx has been frozen for 10 minutes at least now. The issue is that Ryujinx is "Applying RomFS mods" (Console log message) all at the start instead of just as the game needs the file making this situation incredibly slow.
![image](https://github.com/Ryujinx/Ryujinx/assets/25501603/465bfe76-b063-4984-ae0d-76e06c9da228)
![image](https://github.com/Ryujinx/Ryujinx/assets/25501603/22bb32f1-4594-43cc-9c66-023398044053)

### Reproduction steps

1. Make a mod with thousands of RomFS changes
2. Attempt to load the mod

### Log file

```00:00:00.001 |I| Configuration LogValueChange: EnableFileLog set to: True
00:00:00.025 |I| Configuration LogValueChange: ResScale set to: 1
00:00:00.028 |I| Configuration LogValueChange: ResScaleCustom set to: 1
00:00:00.028 |I| Configuration LogValueChange: MaxAnisotropy set to: 2
00:00:00.028 |I| Configuration LogValueChange: AspectRatio set to: Stretched
00:00:00.029 |I| Configuration LogValueChange: BackendThreading set to: Auto
00:00:00.029 |I| Configuration LogValueChange: GraphicsBackend set to: Vulkan
00:00:00.029 |I| Configuration LogValueChange: PreferredGpu set to: 0x10DE_0x1EC7
00:00:00.030 |I| Configuration LogValueChange: AntiAliasing set to: None
00:00:00.030 |I| Configuration LogValueChange: ScalingFilter set to: Bilinear
00:00:00.030 |I| Configuration LogValueChange: ScalingFilterLevel set to: 80
00:00:00.031 |I| Configuration LogValueChange: EnableDockedMode set to: True
00:00:00.035 |I| Configuration LogValueChange: EnableVsync set to: True
00:00:00.035 |I| Configuration LogValueChange: EnableShaderCache set to: True
00:00:00.035 |I| Configuration LogValueChange: EnableTextureRecompression set to: False
00:00:00.035 |I| Configuration LogValueChange: EnableMacroHLE set to: True
00:00:00.035 |I| Configuration LogValueChange: EnableColorSpacePassthrough set to: False
00:00:00.035 |I| Configuration LogValueChange: EnablePtc set to: True
00:00:00.035 |I| Configuration LogValueChange: EnableInternetAccess set to: False
00:00:00.035 |I| Configuration LogValueChange: EnableFsIntegrityChecks set to: True
00:00:00.035 |I| Configuration LogValueChange: FsGlobalAccessLogMode set to: 0
00:00:00.035 |I| Configuration LogValueChange: AudioBackend set to: SDL2
00:00:00.036 |I| Configuration LogValueChange: AudioVolume set to: 1
00:00:00.036 |I| Configuration LogValueChange: MemoryManagerMode set to: HostMappedUnsafe
00:00:00.036 |I| Configuration LogValueChange: ExpandRam set to: False
00:00:00.036 |I| Configuration LogValueChange: IgnoreMissingServices set to: True
00:00:00.036 |I| Configuration LogValueChange: UseHypervisor set to: True
00:00:00.040 |N| Application PrintSystemInfo: Ryujinx Version: 1.1.0+7316df3
00:00:00.044 |N| Application Print: Operating System: Microsoft Windows 10.0.22621 (X64)
00:00:00.044 |N| Application Print: CPU: 13th Gen Intel(R) Core(TM) i9-13900KS ; 32 logical
00:00:00.045 |N| Application Print: RAM: Total 32536 MiB ; Available 15319 MiB
00:00:00.046 |N| Application PrintSystemInfo: Logs Enabled: Info, Warning, Error, Guest, Stub
00:00:00.046 |N| Application PrintSystemInfo: Launch Mode: UserProfile
00:00:00.682 |W| Application CheckScreenBounds: Failed to find valid start-up coordinates. Defaulting to primary monitor center.
00:00:00.091 |I| Gpu : Backend Threading (Auto): True
00:00:00.270 |N| Application LoadGuestApplication: Using Firmware Version: 14.1.2
00:00:00.270 |I| Application LoadGuestApplication: Loading as XCI.
00:00:00.288 |I| ModLoader QueryContentsDir: Searching mods for Title 0100A3D008C5C000
00:00:00.289 |I| ModLoader AddModsFromDirectory: Found mod '0100a3d008c5c000' [R]
00:00:00.289 |I| ModLoader QueryContentsDir: Searching mods for Title 0100A3D008C5C000
00:00:00.291 |I| Loader Load: Loading rtld...
00:00:00.297 |I| Loader PrintRoSectionInfo: rtld:
    Module: nnrtld
00:00:00.297 |I| Loader Load: Loading main...
00:00:00.575 |I| Loader PrintRoSectionInfo: main:
    Module: C:\jenkins\workspace\titan\v_1_0_0\rom_build\programs\prog\app\msvc\main_app\NX64\Release\main_app.nss
    SDK Libraries: SDK MW+Nintendo+NintendoSDK_libcurl-13_3_5-Release
                   SDK MW+Nintendo+NintendoSDK_gRPC_For_NPLN-0_0_0-Release
                   SDK MW+Nintendo+NintendoSDK_OpenSSL_For_NPLN-0_0_0-Release
                   SDK MW+Nintendo+NintendoSDK_NPLN-1_4_1-Release
                   SDK MW+Nintendo+NintendoSDK_Google_APIs_For_NPLN-1_4_1-Release
                   SDK MW+Nintendo+PiaCommon-6_20_1
                   SDK MW+Nintendo+Pia-6_20_1
                   SDK MW+Nintendo+PiaLocal-6_20_1
                   SDK MW+Nintendo+PiaLan-6_20_1
                   SDK MW+Nintendo+PiaReckoning-6_20_1
                   SDK MW+Nintendo+PiaSession-6_20_1
                   SDK MW+Nintendo+PiaTransport-6_20_1
                   SDK MW+Nintendo+PiaNpln-6_20_1
                   SDK MW+NINTENDO+nlib_LZ4-22_0201
                   SDK MW+Nintendo+NintendoWare_Font-13_3_5-Release
                   SDK MW+Nintendo+NintendoWare_Ui2d-13_3_5-Release
                   SDK MW+Nintendo+NintendoWare_Vfx-13_3_5-Release
                   SDK MW+Nintendo+NintendoWare_G3d-13_3_5-Release
                   SDK MW+Nintendo+NintendoSDK_gfx-13_3_5-Release
                   SDK MW+Microsoft+Havok_AI-2021_1
00:00:00.575 |I| Loader Load: Loading subsdk0...
00:00:00.596 |I| Loader PrintRoSectionInfo: subsdk0:
    Module: multimedia
    SDK Libraries: SDK MW+Nintendo+NintendoSDK_movie-13_3_5-Release
00:00:00.596 |I| Loader Load: Loading sdk...
00:00:00.645 |I| Loader PrintRoSectionInfo: sdk:
    Module: nnSdk
    FS SDK Version: 13.3.5
    SDK Libraries: SDK MW+Nintendo+NintendoSDK_libz-13_3_5-Release
                   SDK MW+Nintendo+NintendoSdk_nnSdk-13_3_5-Release
                   SDK MW+Nintendo+NintendoSdk_BuildTarget-64Bit
                   SDK MW+Nintendo+NintendoSDK_NVN-13_3_5-Release
00:00:00.729 |I| Ptc Initialize: Initializing Profiled Persistent Translation Cache (enabled: True).
00:00:03.079 |I| Ptc Load: Loaded Backup Translation Cache (size: 168016072 bytes, translated functions: 81640).
00:00:03.092 |I| Ptc Load: Loaded Profiling Info (size: 256563 bytes, profiled functions: 78304).
00:00:03.098 |I| Loader LoadNsos: Loading image 0 at 0x0000000008500000...
00:00:03.100 |I| Loader LoadNsos: Loading image 1 at 0x0000000008504000...
00:00:03.122 |I| Loader LoadNsos: Loading image 2 at 0x000000000c8f9000...
00:00:03.124 |I| Loader LoadNsos: Loading image 3 at 0x000000000cf93000...
00:00:03.129 |I| ModLoader LoadCheats: Build ids found for title 0100A3D008C5C000:
    567CDD4C5251494C5CD70822E3CCC740C190C2B2000000000000000000000000
    CA44E7DD62E61A2A4E8889B3319AE173375770FD000000000000000000000000
    D0F13015E5BADF2243D60A2BC9B978DDDBA80608000000000000000000000000
    3644A6760469451CABC99F00C540ACC2A64AEE41000000000000000000000000
00:00:03.131 |I| ModLoader ApplyRomFsMods: Applying RomFS mods for Title 0100A3D008C5C000```

### OS

Windows 11

### Ryujinx version

1.1.0+7316df3 (Latest as of issue with ava ui)

### Game version

1.0.0

### CPU

i9-13900ks

### GPU

NVIDIA RTX 2070 SUPER

### RAM

32GB

### List of applied mods

Custom mod which extracts the internal archive and invalidates the archives entries to make file access faster

### Additional context?

_No response_

**Comment by MutantAura** [comment]
This is simply how layeredFS mods work? The more you replace, the more time it going to be spent reconstructing the romfs? It's like asking why copying 10TB of data takes a long time when copying 1mb doesn't?

**Comment by hYdos** [comment]
that's the issue. Layered FS could easily get the files just in time and
avoid this issue.

**Comment by OldManKain** [comment]
This is how LayeredFS mods work, if you would like them to load faster you would need to convert your loose files to a romfs.bin (not sure if Ryujinx is compatible with those).   I'll do some testing.

**Comment by hYdos** [comment]
Its an issue with the implementation of LayeredFS not the idea of LayeredFS mods itself

**Comment by OldManKain** [comment]
Which means it's not a Ryujinx problem.  It's a fs.mitm problem.

**Comment by hYdos** [comment]
Does Ryujinx not implement LayeredFS?

**Comment by OldManKain** [comment]
Ryujinx ha  layeredFS compatibility based off Atmosphere, it works just like it would on real hardware.

**Comment by hYdos** [comment]
Did some new research, Atmosphere caches the result after building but Ryujinx does not meaning consecutive runs will run poorly as can be seen here

https://github.com/Ryujinx/Ryujinx/assets/25501603/0bf8880e-6be6-45ea-82f5-042379121720

**Comment by SciresM** [comment]
@OldManKain @MutantAura some casual looking through the code strongly suggests this is a problem with Ryu's impl.

**Comment by SciresM** [comment]
I would re-open this issue, which appears to me to be real. There is no reason for romfs metadata build to take ten minutes; it takes O(10-20 seconds) on hardware, even with hundreds of thousands of files.

**Comment by SciresM** [comment]
I mean, there are games with a quarter million files. Not sure why the steps taken on hardware wouldn't work?

**Comment by hYdos** [comment]
![image](https://github.com/Ryujinx/Ryujinx/assets/25501603/36da8c7d-af8e-4aa1-93da-0e255dd1fc43)
here are my results in java from getting all files in that folder. 10 seconds

**Comment by AcK77** [comment]
Superseeded by #6180

**Comment by TSRBerry** [comment]
Are we sure this issue is fixed with #6180?

**Comment by AcK77** [comment]
It's fixed in https://github.com/Ryujinx/Ryujinx/pull/5672 and it's the same code...
